### PR TITLE
Scrolling2.0

### PR DESCRIPTION
**Pull request by KirbyG**
I factored out the rendering code into an artist class instantiated by np-gravity. Now np-gravity deals exclusively with reading user input and sending it where it needs to go.

### CODE REVIEW

File: artist.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        self.game = game
+
+        # pygame setup
+        os.environ['SDL_VIDEO_WINDOW_POS'] = "%d,%d" % (0,0) # locate the window in the upper left
+        pygame.init()
+        pygame.display.set_caption(f'{type(game).__name__} Reduction')
+
+        # compute constants
+
+        # this is the approximate height of the menu bar at the top of iOS
```
Comment (JacobLezberg) [inline artist.py:17]: Why is iOS relevant?

File: artist.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        # how big could each block dimension be if we used the whole display
+        FITTED_BLOCK_DIM = SCREEN_DIM / game.grid.dim
+
+        MIN_BLOCK_DIM = 15 # arbitrary cutoff for ease of visualization
+
+        # blocks need to be square and not smaller than the min
+        self.BLOCK_DIM = max(MIN_BLOCK_DIM, min(FITTED_BLOCK_DIM()))
+
+        # limit the max window dimension to fit on the display, and ensure it multiplies the block dimension
+        self.WINDOW_DIM = Vector(int(round(min(SCREEN_DIM.x - (SCREEN_DIM.x % self.BLOCK_DIM), self.BLOCK_DIM * game.grid.dim.x))), int(round(min(SCREEN_DIM.y - (SCREEN_DIM.y % self.BLOCK_DIM), self.BLOCK_DIM * game.grid.dim.y))))
```
Comment (JacobLezberg) [inline artist.py]: This is really hard to parse when it's nested. Feel free to rename vars if inaccurate. Also you don't need the `int()` call because `round()` without a second param creates an `int`.
```suggestion
        window_width = min(SCREEN_DIM.x - (SCREEN_DIM.x % self.BLOCK_DIM), self.BLOCK_DIM * game.grid.dim.x)
        window_height = min(SCREEN_DIM.y - (SCREEN_DIM.y % self.BLOCK_DIM), self.BLOCK_DIM * game.grid.dim.y)
        self.WINDOW_DIM = Vector(round(window_width), round(window_height))
```

File: artist.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        
+        # the number of blocks that will fit in a window at a time
+        self.WINDOW_BLOCKS = Vector(int(round(self.WINDOW_DIM.x / self.BLOCK_DIM)), int(round(self.WINDOW_DIM.y / self.BLOCK_DIM)))
```
Comment (JacobLezberg) [inline artist.py]:
```suggestion
        self.WINDOW_BLOCKS = Vector(round(self.WINDOW_DIM.x / self.BLOCK_DIM), round(self.WINDOW_DIM.y / self.BLOCK_DIM))
```

File: artist.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        self.draw()
+
+    # all params are in block units
+    # return format is [corner x, corner y, width, height]
+    
+    # note that the y coordinate is inverted because pygame considers the upper
+    # left corner of the window to be the origin, while all logical objects
+    # rely on a more standard coordinate system originating in the lower left
+    def grid_to_px(self, x, y, offset):
+        return [(x - offset.x) * self.BLOCK_DIM + 1, (self.WINDOW_BLOCKS.y - 1 - (y - offset.y)) * self.BLOCK_DIM + 1, self.BLOCK_DIM - 1, self.BLOCK_DIM - 1]
```
Comment (JacobLezberg) [inline artist.py]: Don't need the comment if we just label these as variables
```suggestion
    # all params are in block units   
    # Note: the y coordinate is inverted because pygame considers the upper
    # left corner of the window to be the origin, while all logical objects
    # rely on a more standard coordinate system originating in the lower left
    def grid_to_px(self, x, y, offset):
        corner_x = (x - offset.x) * self.BLOCK_DIM + 1
        corner_y = (self.WINDOW_BLOCKS.y - 1 - (y - offset.y)) * self.BLOCK_DIM + 1
        width = height = self.BLOCK_DIM - 1
        return [corner_x, corner_y, width, height]
```

File: artist.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+    def draw(self):
+        # compute the offset vector
+        offset = Vector(min(self.game.grid.dim.x - self.WINDOW_BLOCKS.x, max(0, self.game.player.pos.x - self.CENTER.x)), min(self.game.grid.dim.y - self.WINDOW_BLOCKS.y, max(0, self.game.player.pos.y - self.CENTER.y)))
+        
+        # set the background to black
+        pygame.draw.rect(self.screen, (0, 0 ,0), [0, 0, self.WINDOW_DIM.x, self.WINDOW_DIM.y])
```
Comment (JacobLezberg) [inline artist.py]: Do you think we should add `background: (0, 0, 0)` as a color in the constants file and use it here?

File: artist.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    
+    # note that the y coordinate is inverted because pygame considers the upper
+    # left corner of the window to be the origin, while all logical objects
+    # rely on a more standard coordinate system originating in the lower left
+    def grid_to_px(self, x, y, offset):
+        return [(x - offset.x) * self.BLOCK_DIM + 1, (self.WINDOW_BLOCKS.y - 1 - (y - offset.y)) * self.BLOCK_DIM + 1, self.BLOCK_DIM - 1, self.BLOCK_DIM - 1]
+
+    def draw(self):
+        # compute the offset vector
+        offset = Vector(min(self.game.grid.dim.x - self.WINDOW_BLOCKS.x, max(0, self.game.player.pos.x - self.CENTER.x)), min(self.game.grid.dim.y - self.WINDOW_BLOCKS.y, max(0, self.game.player.pos.y - self.CENTER.y)))
```
Comment (JacobLezberg) [inline artist.py]: Same thing, I think it will be more readable separated (and you can see how the axes mirror each other which is nice)
```suggestion
        x_offset = min(self.game.grid.dim.x - self.WINDOW_BLOCKS.x, max(0, self.game.player.pos.x - self.CENTER.x))
        y_offset = min(self.game.grid.dim.y - self.WINDOW_BLOCKS.y, max(0, self.game.player.pos.y - self.CENTER.y))
        offset = Vector(x_offset, y_offset)
```

File: artist.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+# in charge of drawing Game objects
+class Artist:
+    def __init__(self, game):
+        self.game = game
+
+        # pygame setup
+        os.environ['SDL_VIDEO_WINDOW_POS'] = "%d,%d" % (0,0) # locate the window in the upper left
+        pygame.init()
+        pygame.display.set_caption(f'{type(game).__name__} Reduction')
```
Comment (JacobLezberg) [inline artist.py:13]: Not sure if this is the right place for this...is Artist just for rendering the contents of the screen? And we're leaving NP-Gravity to handle the "metagame" (i.e. pygame init, game loop, window title, stuff like that)?

File: megalit.py [hunk 8]
```diff
@@ -36,9 +36,11 @@ def reduce(self):
 
         # container for all slabs in the level
         self.slabs = []
-
+        # level for demoing bi-driectional scrolling
```
Comment (JacobLezberg) [inline megalit.py]:
```suggestion
        # level for demoing bi-directional scrolling
```

**Review by JacobLezberg — CHANGES_REQUESTED** [review]

File: artist.py (hunk 1 above)
Comment (KirbyG) [reply artist.py:17]: on my mac i had to add this buffer because the dimensions of my display returned by pygame did not exactly correlate to the actual usable height of my display because of the menu bar at the top

File: artist.py (hunk 7 above)
Comment (KirbyG) [reply artist.py:13]: i think this is ok as the os.environ and set_caption calls are graphics related and i'm conceptualizing Artist as handling everything related to graphics output, while np-gravity will specialize in taking user input and sending it where it needs to go. the pygame.init() call is debatable, but since it would be the only pygame call in np-gravity and we do want to ensure pygame has been initialized before we start making calls to it in Artist, i'm in favor of leaving it here

**Review by JacobLezberg — APPROVED** [review]
